design/baremetal-operator/user-defined-parameters.md

```diff
@@ -0,0 +1,456 @@
+        if err != nil {
+                return actionError{err}
+@@ -796,7 +804,7 @@ func (r *BareMetalHostReconciler) registerHost(prov provisioner.Provisioner, inf
+                        PreprovisioningImage:  preprovImg,
+                        HasCustomDeploy:       hasCustomDeploy(info.host),
+                },
+-               credsChanged,
++               credsChanged || extraParametersChanged,
+                info.host.Status.ErrorType == metal3v1alpha1.RegistrationError)
+
+        if errors.Is(err, provisioner.ErrNeedsPreprovisioningImage) &&
+@@ -854,7 +862,7 @@ func (r *BareMetalHostReconciler) registerHost(prov provisioner.Provisioner, inf
+                }
+        }
+
+-       // Reaching this point means the credentials are valid and worked,
++       // Reaching this point means the credentials/extraParameters are valid and worked,
+        // so clear any previous error and record the success in the
+        // status block.
+        registeredNewCreds := !info.host.Status.GoodCredentials.Match(*info.bmcCredsSecret)
+@@ -863,11 +871,21 @@ func (r *BareMetalHostReconciler) registerHost(prov provisioner.Provisioner, inf
+                info.host.UpdateGoodCredentials(*info.bmcCredsSecret)
+                info.publishEvent("BMCAccessValidated", "Verified access to BMC")
+                dirty = true
+-       } else {
++       }
++
++       registeredNewExtraParameters := !info.host.Status.GoodExtraParameters.Match(*info.bmcExtraParametersSecret)
++       if registeredNewExtraParameters {
++               info.log.Info("updating extra parameters success status fields")
++               info.host.UpdateGoodExtraParameters(*info.bmcExtraParametersSecret)
++               info.publishEvent("BMCExtraParametersValidated", "Verified access to BMC")
++               dirty = true
++       }
++
++       if !registeredNewCreds && !registeredNewExtraParameters {
+                info.log.Info("verified access to the BMC")
+        }
+
+-       if info.host.Status.ErrorType == metal3v1alpha1.RegistrationError || registeredNewCreds {
++       if info.host.Status.ErrorType == metal3v1alpha1.RegistrationError || registeredNewCreds || registeredNewExtraParameters {
+                info.log.Info("clearing previous error message")
+                dirty = clearError(info.host)
+        }
+```
+
+### Risks and Mitigations
+
+None
+
+### Work Items
+
+- Add a new field `ExtraParametersName`.
+- Convert `ExtraParametersName` to `ExtraParameters`.
+- Modify provisioner factory interface to accept extra parameters.
+- Modify ironic provisioner to merge the extra parameters to driverinfo.
+- Update the `ExtraParameters` to ironic when they change
+
+### Dependencies
+
+None
+
+### Test Plan
+
+- Unit tests
+- Integration testing with iRMC hardware
+
+### Upgrade / Downgrade Strategy
+
+None
+
+### Version Skew Strategy
+
+None
+
+## Drawbacks
+
+Due to the limitation of `Secret` object, the value of `ExtraParameter` can only
+be string type.
+
```

Comment:
The proposal gives users no way to determine which fields can be used without diving into the Ironic docs.

Reply:
This is indeed a drawback, and we will try to improve it by putting relevant documentation links in the api description. We'll  alos add this to Drawbacks.